tempto-core/src/main/java/com/teradata/tempto/s3/S3Client.java

```diff
@@ -0,0 +1,232 @@
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import com.amazonaws.services.s3.model.ListObjectsV2Result;
+import com.google.common.base.Splitter;
+import com.google.common.collect.Ordering;
+import com.google.inject.name.Named;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * S3 client based on the Amazon AWS client API
+ */
+public class S3Client
+{
+    private static final Logger logger = getLogger(S3Client.class);
+    private static final String PATH_SEPARATOR = "/";
+    private static final Splitter PATH_SPLITTER = Splitter.on('/').omitEmptyStrings();
+
+    public static final String CONF_S3_AUTH_KEY_ID = "s3.auth.keyid";
+    public static final String CONF_S3_AUTH_SECRET = "s3.auth.secret";
+    public static final String CONF_S3_BUCKET_NAME = "s3.bucketname";
+
+    private final String bucketName;
+    private final AmazonS3 client;
+
+    public S3Client(
+            @Named(CONF_S3_AUTH_KEY_ID) String keyId,
+            @Named(CONF_S3_AUTH_SECRET) String secret,
+            @Named(CONF_S3_BUCKET_NAME) String bucketName)
+    {
+        // Check null values and set object variables
+        checkNotNull(keyId, "aws api access key id is null");
+        checkNotNull(secret, "aws api secret access key is null");
+        this.bucketName = checkNotNull(bucketName, "s3 bucket name is null");
+
+        // Create S3 client
+        client = AmazonS3ClientBuilder.standard()
+                .withCredentials(new AWSStaticCredentialsProvider(buildAWSCredentials(keyId, secret)))
+                .build();
+    }
+
+    private BasicAWSCredentials buildAWSCredentials(String keyId, String secret)
+    {
+        try {
+            return new BasicAWSCredentials(keyId, secret);
+        } catch (Exception e) {
+            throw new AmazonClientException("Cannot initialize credentials", e);
+        }
+    }
+
+    public boolean exists(String path)
+    {
+        return client.doesObjectExist(bucketName, path);
+    }
+
+    public List<String> listFiles(String path)
+    {
+        // Get list of all files located in this path
+        ListObjectsV2Result listing = client.listObjectsV2(new ListObjectsV2Request()
+                .withBucketName(bucketName)
+                .withPrefix(ensureTrailingSlashPresent(path))
+                .withDelimiter(PATH_SEPARATOR));
+
+        // Return all objects (filtering out the path itself)
+        return listing.getObjectSummaries().stream()
+                .map(S3ObjectSummary::getKey)
+                .filter(key -> !key.endsWith(PATH_SEPARATOR))
+                .collect(Collectors.toList());
+    }
+
+    private String ensureTrailingSlashPresent(String path)
+    {
+        if (!path.endsWith(PATH_SEPARATOR)) {
+            path = path + PATH_SEPARATOR;
+        }
+        return path;
+    }
+
+    public List<String> listFolders(String path)
```

Comment:
The getCommonPrefixes() method of the ListObjectsV2Result object returns all folders within the passed in prefix.  So for example if you had a directory structure like,

foo/bar/baz
foo/bar/bash
foo/boo

It will only return "foo/bar" since it is the only folder (common prefix) listed.

Modified the comment to say "Return all folders located in this path"